source/index.ts

```diff
@@ -79,13 +81,216 @@ export class LockAcquisitionError extends Error {
+			}
+
+			amount -= release;
+			reply.reply({
+				type: MessageType.SEMAPHORE_RELEASE,
+				amount: release
+			});
+		};
+	}
+}
+
+export class UnmanagedSemaphore {
+	readonly #context: SharedContext;
+
+	constructor(
+		context: SharedContext,
+		public readonly id: string,
+		public readonly initialValue: number
+	) {
+		if (initialValue < 0 || !Number.isSafeInteger(initialValue)) {
+			throw new RangeError('initialValue must be a non-negative safe integer');
+		}
+
+		this.#context = context;
+	}
+
+	async down(amount = 1) {
+		if (amount < 0 || !Number.isSafeInteger(amount)) {
+			throw new RangeError('amount must be a non-negative safe integer');
+		}
+
+		// Allow down() to be called before the shared worker is availabe.
+		await protocol.available;
+
+		await downSemaphore(this, this.#context.id, amount, true);
+	}
+
+	async downNow(amount = 1) {
+		if (amount < 0 || !Number.isSafeInteger(amount)) {
+			throw new RangeError('amount must be a non-negative safe integer');
+		}
+
+		// Down immediately, which will fail if the protocol is not available.
+		// "Now" should not mean "wait until we're ready."
+
+		await downSemaphore(this, this.#context.id, amount, false);
+	}
+
+	async up(amount = 1) {
+		if (amount < 0 || !Number.isSafeInteger(amount)) {
+			throw new RangeError('amount must be a non-negative safe integer');
+		}
+
+		// Allow up() to be called before the shared worker is availabe.
+		await protocol.available;
+
+		const {id, initialValue} = this;
+		const message = protocol.publish({
+			type: MessageType.SEMAPHORE_UP,
+			contextId: this.#context.id,
+			semaphore: {managed: false, id, initialValue},
+			amount
+		});
+
+		for await (const reply of message.replies()) {
+			if (reply.data.type === MessageType.SEMAPHORE_SUCCEEDED) {
+				return;
+			}
+
+			if (reply.data.type === MessageType.SEMAPHORE_CREATION_FAILED) {
+				throw new SemaphoreCreationError(this, reply.data);
+			}
+		}
+
+		/* c8 ignore next 2 */
+		// The above loop will never actually break if the resources are not acquired.
+		return never();
+	}
+}
+
+type Semaphore = ManagedSemaphore | UnmanagedSemaphore;
+
+async function downSemaphore(semaphore: Semaphore, contextId: string, amount: number, wait: boolean): Promise<ReceivedMessage> {
+	const {id, initialValue} = semaphore;
+	const message = protocol.publish({
+		type: MessageType.SEMAPHORE_DOWN,
+		contextId,
+		semaphore: {managed: semaphore instanceof ManagedSemaphore, id, initialValue},
+		amount,
+		wait
+	});
+
+	for await (const reply of message.replies()) {
+		if (reply.data.type === MessageType.SEMAPHORE_SUCCEEDED) {
+			return reply;
+		}
+
+		if (reply.data.type === MessageType.SEMAPHORE_FAILED) {
+			throw new SemaphoreDownError(id, amount);
+		}
+
+		if (reply.data.type === MessageType.SEMAPHORE_CREATION_FAILED) {
+			throw new SemaphoreCreationError(semaphore, reply.data);
+		}
+	}
+
+	/* c8 ignore next 2 */
```

Comment:
Should these ignore comments perhaps be removed? I couldn't find a good way to satisfy both c8 and eslint:
- With the `/* c8 ignore next 2 */` comments are currently used, c8 just warns about the blank line before and the `}` line after not being covered.
- Removing the line above and increasing the range to `/* c8 ignore next 3 */ makes c8 happy, but eslint says "Expected blank line before this statement."
- Moving the comment up to line 237 and increasing the range to `/* c8 ignore next 4 */ works, but doesn't exactly look great.

Reply:
Ugh yea, getting rid.